Mathlib/Analysis/SpecialFunctions/BinEntropy.lean

```diff
@@ -0,0 +1,710 @@
+
+lemma binaryEntropy_eq_log2_iff_eq_half {p : ℝ} (pge0 : 0 ≤ p) (ple1 : p ≤ 1) :
+    binaryEntropy p = log 2 ↔ p = 1/2 := by
+  constructor <;> intro h
+  · by_cases h' : p < 1/2
+    · linarith [binaryEntropy_lt_log2_of_lt_half pge0 h']
+    · by_cases pgthalf : 1/2 < p
+      · linarith [binaryEntropy_lt_log2_of_gt_half pgthalf ple1]
+      · linarith
+  · simp only [one_div, binaryEntropy_onehalf, h]
+
+lemma binaryEntropy_le_log2 {p : ℝ} (pge0 : 0 ≤ p) (ple1 : p ≤ 1) :
+    binaryEntropy p ≤ log 2 := by
+  by_cases hh: p = 1/2
+  · simp only [one_div, binaryEntropy_onehalf, le_refl, hh]
+  · by_cases gg: binaryEntropy p = log 2
+    · simp only [le_refl, gg]
+    · by_cases hhh: p < 1/2
+      · linarith [binaryEntropy_lt_log2_of_lt_half pge0 hhh]
+      · have : 1/2 < p := by
+          refine Ne.lt_of_le ?_ ?_
+          · intro
+            simp_all only [not_lt, not_true_eq_false]
+          · simp_all only [one_div, not_lt]
+        exact (binaryEntropy_lt_log2_of_gt_half this ple1).le
+
+/-- The q-ary entropy function is continuous everywhere.
+This is due to definition of `Real.log` for negative numbers. -/
+@[fun_prop] lemma qaryEntropy_continuous {q : ℕ} : Continuous (qaryEntropy q) := by
+  refine Continuous.add ?_ (Continuous.neg ?_)
+  · exact Continuous.sub (by fun_prop) continuous_mul_log
+  · exact Continuous.comp continuous_mul_log (continuous_sub_left 1)
+
+/-- Binary entropy is continuous everywhere.
+This is due to definition of `Real.log` for negative numbers. -/
+@[fun_prop] lemma binaryEntropy_continuous : Continuous binaryEntropy := qaryEntropy_continuous
+
+/-! ### Derivatives of binary entropy function -/
+
+@[simp] lemma deriv_one_minus (x : ℝ) : deriv (fun (y : ℝ) ↦ 1 - y) x = -1 := by
+  have onem (y : ℝ) : 1 - y = -(y + -1) := by ring
+  simp only [onem, neg_add_rev, neg_neg, differentiableAt_const, deriv_const_add', deriv_neg'']
+
+section general
+
+variable {𝕜 : Type*} [NontriviallyNormedField 𝕜]
+variable {E : Type*} [NormedAddCommGroup E] [NormedSpace 𝕜 E]
+variable {F : Type*} [NormedAddCommGroup F] [NormedSpace 𝕜 F]
+variable {f g : E → F}
+variable {x : E}
+
+lemma DifferentiableAt.add_iff_left (hg : DifferentiableAt 𝕜 g x) :
+    DifferentiableAt 𝕜 (fun y => f y + g y) x ↔ DifferentiableAt 𝕜 f x := by
+  constructor <;> intro h
+  · have f_eq_sum_sub_g: f = (fun y => f y + g y) - g := by
+      ext
+      simp only [Pi.sub_apply, add_sub_cancel_right]
+    rw [f_eq_sum_sub_g]
+    exact DifferentiableAt.sub h hg
+  · simp_all only [DifferentiableAt.add]
+
+lemma DifferentiableAt.add_iff_right (hg : DifferentiableAt 𝕜 f x) :
+    DifferentiableAt 𝕜 (fun y => f y + g y) x ↔ DifferentiableAt 𝕜 g x := by
+  rw [show (fun y ↦ f y + g y) = (fun y ↦ g y + f y) by ext; rw [add_comm]]
+  exact hg.add_iff_left
+
+lemma differentiableAt_iff_differentiableAt_comp_mul_add
+    {a b m : 𝕜} (hm : m ≠ 0) (f : 𝕜 → E) :
+    DifferentiableAt 𝕜 f a ↔ DifferentiableAt 𝕜 (fun x => f (m * x + b)) (m⁻¹ * (a - b)):= by
```

Comment:
Is it specific? I would think it could come up often when trying to prove something is *not* differentiable, assuming, e.g., this junk value thing is worth doing. I would maybe rename it to `DifferentiableAt.iff_comp_mul_add`. It seems to me no less useful than `add_iff_left`.

There's also things like several variants of ``[intervalIntegral.mul_integral_comp_mul_add](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/SpecialFunctions/Integrals.html#intervalIntegral.mul_integral_comp_mul_add).